pysph/tools/ipy_viewer.py

```diff
@@ -920,210 +1149,248 @@ def _create_widgets(self):
 
         array_name = change['owner'].owner
         pa_widgets = self._widgets.particles[array_name]
-        change['new'] = pa_widgets.scalar.value
-        self._scalar_handler(change)
-        self._legend_handler(None)
+        if pa_widgets.is_visible.value is True:
+            change['new'] = pa_widgets.scalar.value
+            self._scalar_handler(change)
 
     def _legend_handler(self, change):
 
-        temp_data = self.get_frame(self._widgets.frame.value)
+        temp_data = self.get_frame(self._widgets.frame.value)['arrays']
         with self.legend:
-            self.pltfigure.clf()
-            clear_output(wait=True)
-            self.cbars = {}
-            self.cbars_ax = {}
+            for _cbar_ax in self._cbar_ax.values():
+                _cbar_ax.clear()
+                self.pltfigure.delaxes(_cbar_ax)
+            self._cbar_ax = {}
+            self._cbars = {}
             for array_name in self._widgets.particles.keys():
                 pa_widgets = self._widgets.particles[array_name]
-                if pa_widgets.scalar.value != 'None':
-                    cmap = getattr(mpl.cm, pa_widgets.scalar_cmap.value)
-                    ticks = set(list(np.sort(
-                        getattr(
-                            temp_data['arrays'][array_name],
+                if (pa_widgets.scalar.value != 'None' and
+                        pa_widgets.legend.value is True):
+                    if pa_widgets.is_visible.value is True:
+                        cmap = getattr(mpl.cm, pa_widgets.scalar_cmap.value)
+                        c = getattr(
+                            temp_data[array_name],
                             pa_widgets.scalar.value
                         )
-                    )))
-                    ticks = list(ticks)
-                    ticks.sort()
-                    if len(ticks) == 1:
-                        ticks.append(ticks[0] + 0.00000001)
-                        # To avoid passing a singleton set
-
-                    self.cbars_ax[array_name] = self.pltfigure.add_axes(
-                        [
-                            0.2*len(self.cbars_ax.keys()),
-                            0,
-                            0.05,
-                            0.5
-                        ]
-                    )
-                    self.cbars[array_name] = mpl.colorbar.ColorbarBase(
-                        ax=self.cbars_ax[array_name],
-                        cmap=cmap,
-                        boundaries=ticks,
-                    )
-                    self.cbars[array_name].set_label(
-                                    array_name+" : "+pa_widgets.scalar.value
-                            )
-            display(self.pltfigure)
+                        arr = self._normalized_for_colormaps(c)
+                        minm = arr[0]
+                        maxm = arr[1]
+                        absmaxm = arr[2]
+
+                        # Do not change the order of the if-else #
```

Comment:
Seems like a lot of copy-paste code is here, don't repeat code, try to write the code so you write things only once and reuse things as much as possible.  You can do this with a subclass for example.

Reply:
I've implemented most of your suggestions. However in certain functions like the `_legend_handler`, even though both the `3DViewer` and the `2DViewer` share some code, there are kinks (like changing the axes positions before specific events/to specific positions) that kind of make it tough to write reusable code.

Reply:
It may be difficult but you should try very hard to see if they can be written once, for example if it is just an option you can generalize the function to handle both cases.

Reply:
Of course, not at the expense of things being readable and sane.